RecoBTag/SoftLepton/plugins/SoftPFMuonTagInfoProducer.cc

```diff
@@ -26,132 +33,127 @@
+  const reco::Vertex* vertex=&theVertexCollection->front();
+  
+  // Biult TransientTrackBuilder
+  edm::ESHandle<TransientTrackBuilder> theTrackBuilder;
+  iSetup.get<TransientTrackRecord>().get("TransientTrackBuilder", theTrackBuilder);
+  const TransientTrackBuilder* transientTrackBuilder=theTrackBuilder.product();
+  
+  // Loop on jets
+  for(unsigned int ij=0, nj=theJetCollection->size(); ij<nj; ij++) {
+    edm::RefToBase<reco::Jet> jetRef = theJetCollection->refAt(ij);
+    // Build TagInfo object
+    reco::CandSoftLeptonTagInfo tagInfo;
+    tagInfo.setJetRef(jetRef);
+    // Loop on jet daughters
+    for(unsigned int id=0, nd=jetRef->numberOfDaughters(); id<nd; ++id) {
+      edm::Ptr<reco::Candidate> lepPtr = jetRef->daughterPtr(id);
+      if(abs(lepPtr->pdgId())!=13) continue;
+      
+      const reco::Muon* muon(NULL);
+      // Step 1: try to access the muon from reco::PFCandidate
+      const reco::PFCandidate* pfcand=dynamic_cast<const reco::PFCandidate*>(lepPtr.get());
+      if(pfcand) {
+        muon=pfcand->muonRef().get();
+      }
+      // If not PFCandidate is available, find a match looping on the muon collection
+      else {
+        for(unsigned int im=0, nm=theMuonCollection->size(); im<nm; ++im) { // --- Begin loop on muons
+          const reco::Muon* recomuon=theMuonCollection->refAt(im).get();
+          const pat::Muon* patmuon=dynamic_cast<const pat::Muon*>(recomuon);
+          // Step 2: try a match between reco::Candidate
+          if(patmuon) {
+            if(patmuon->originalObjectRef()==lepPtr) {
+              muon=theMuonCollection->refAt(im).get();
+              break;
+            }
+          }
+          // Step 3: try a match with dR and dpT if pat::Muon casting fails
+          else {
+            if(deltaR(*recomuon, *lepPtr)<0.01 && fabs(recomuon->pt()-lepPtr->pt())/lepPtr->pt()<0.1) {
+              muon=theMuonCollection->refAt(im).get();
+              break;
+            }
+          }
+        } // --- End loop on muons
+      }
+      if(!muon || !muon::isLooseMuon(*muon) || muon->pt()<pTcut) continue;
+      reco::TrackRef trkRef( muon->innerTrack() );
+      reco::TrackBaseRef trkBaseRef( trkRef );
+      
+      // Build Transient Track
+      reco::TransientTrack transientTrack=transientTrackBuilder->build(trkRef);
+      // Calculate variables
+      reco::SoftLeptonProperties properties;
+      properties.sip2d    = IPTools::signedTransverseImpactParameter(transientTrack, GlobalVector(jetRef->px(), jetRef->py(), jetRef->pz()), *vertex).second.significance();
+      properties.sip3d    = IPTools::signedImpactParameter3D(transientTrack, GlobalVector(jetRef->px(), jetRef->py(), jetRef->pz()), *vertex).second.significance();
+      properties.deltaR   = deltaR(*jetRef, *muon);
+      properties.ptRel    = ( (jetRef->p4().Vect()-muon->p4().Vect()).Cross(muon->p4().Vect()) ).R() / jetRef->p4().Vect().R(); // | (Pj-Pu) X Pu | / | Pj |
```

Comment:
- better extract xxx->p4().Vect() upfront and overate with them instead of copy-pasting multiple times
- ``` - muon->p4().Vect() ``` will just introduce extra numerical uncertainty: why is it here (x.Cross(x) = 0 ?)